assignments/week_2/exercise_2_3/exercise_2_3.md

```diff
@@ -0,0 +1,360 @@
+```
+679 % 16 = 7
+```
+
+Now that is interesting, that is the left most digit of our hex representation.
+Now let us see if we can get the next digit as well, so we divide out the first digit
+by taking the entire number and dividing by 16 (remember we have integers so we will
+not have any decimals).
+
+```
+679 / 16 = 42 // remember integer division lops off the decimals
+```
+
+Now we take out mod again
+
+```
+42 % 16 = 10 = A
+```
+
+Now we can see out algorithm begin to take shape. The summarized steps are
+
+```
+679 % 16 = 7
+679 / 16 = 42
+42 % 16 = 10 = A
+42 / 16 = 2
+2 % 16 = 2
+```
+So if we combine out mod operators we end up with 7A2, now the C++ algorithms header
+gives us a function to reverse a string that can be called like this (make sure to have the header file.
+
+```c++
+#include <algorithm>
+// lotta cool code
+std::string str = "str";
+std::reverse(str.begin(), str.end());
+// str = "rts" now
+```
+
+### Binary To Hex Tips
+Now there is an interesting relationship between binary and hex. Let's use the same
+number as before.
+
+```
+0x2A7 = 7 * 16^0 + 10 * 16^1 + 2*16^2 = 7 + 160 + 512 = 679
+```
+
+Now lets keep in mind that 2^4 = 16 and see if we can use that fact. Let's just look
+at the max value hex can represent with a single digit (15), what about 0xF0?
+
+```
+0xF = 15 = b1111
+0xF0 = 15*16^1 = b11110000
+```
+
+in fact all hex digits can be represented in binary using only 4 values. We can
+use this fact to go from binary to hex. Let us rewrite each digit of hex in binary and see how this
+fits in, we will make sure each digit gets represented with 4 binary digits, we pad with zeros
+
+```
+0x7 = 7 = b0111 = 1*2^0 + 1*2^1 + 1*2^2+ 0*2^3 = 1 + 2 + 4 + 0
+0xA = 10 = b1010 = 0*2^0 + 1*2^1 + 0*2^2 + 1*2^3 = 0 + 2 + 0 + 8
+0x2 = 2 = b0010 = 0*2^0 + 1*2^1 + 0*2^2 + 0*2^3 = 0 + 2 + 0 + 0
+```
+
+Now if we take these and put them in sequence we get the binary representation of
+the original hex number in binary.
+
+```
+0x2A7 = 0010 1010 0111
+```
+
+## Part 3 to Binary
+Now we will be writing two methods that convert from hex and decimal to binary.
+Write two methods with the correct inputs and return type (you should see the pattern by now).
+
+use the previous information to correctly convert from hex and decimal to binary.
+
+## Part 4
+use cin to take in a character that represents what type of number to print out. Using
+the following convention
+
+```
+b = binary
+d = decimal
+h = hex
+```
+
+you will then need to figure out what type of number input you have based on the conventions
+
+```
+0x = hex
+b = binary
+otherwise decimal
+```
+
+the input and output for an example run should match what you see below.
+You can check your output by running the different example.txt files using the command,
+Your output should match
+
+```bash
+cat example4_1.txt | ./exercise_2_3
+input: 2A7
+output type: d
+679
+decimal_inputs: 0
+binary_inputs: 0
+hex_inputs: 1
+decimal_outputs: 1
+binary_outputs: 0
+hex_outputs: 0
+```
```

Comment:
This isn't really the output they'll see though? They won't see the input from `example4_1.txt`.

Reply:
Also, 4_1???

Reply:
part 4 example 1. I mean this is the output they should make it do since that is what we will check in github classroom?